### ISSUE TITLE
viability of a 2 node cluster?

### ISSUE THREAD

**Issue by syadnom** [lizardfs/lizardfs#430]
I've built a 2.5 node proxmox cluster.  2 VM hosts and a third small node that only handles quorum.  I can replace the third node once I need more compute/ram.

The cluster currently runs glusterfs to mirror local disk directories to enable live migration and some redundancy.  This works fine, but glusterfs has some limitations and performance issues in this type of configuration.  A big limitation is that having redundancy requires matched storage nodes, so to expand I have to go from 2 nodes to 4 nodes.

Lizardfs is something I'm looking to for storage for another project so going to 1 storage model makes sense.

So the question is, how viable is a 2 node lizardfs system with the master and shadow on the two 'main' nodes, as well as local storage?  

A related question, can lizardfs present as iscsi or nfs storage?  To utilize a filesystem for live migration, I either need support in proxmox (so it knows the storage is available from other nodes) or to by a dns addressed nfs or iscsi share.

Thanks.

**Comment by 4Dolio** [comment]
I have recently begun to toy around with LFS backed nfs or iscsi for XenServer to use as an SR(Storage Resource).  I use corosync+pacemaker for the floating IP and failover.  NFS did not work so well during a failover transition, but iscsi seems to work great.

As an added kick, the LFS masters+chunkservers are VMs within the Xen Pool, One VM per Xen Host, and those VMs leverage the local storage available on each Xen Host.  I'm not saying this is a good idea, but it appears that it might work, maybe not well but perhaps well enough.

I just added a Corosync Setup and Initialize script to my pull request that might be helpful.
https://github.com/4Dolio/lizardfs/tree/4Dolio-corosync-ocf-metadataserver-patch-1/src/ha-cluster

**Comment by 4Dolio** [comment]
While I do not thing corosync+pacemaker are the best, especially with some of the safeties off as I am using it is a great idea, but it has been working in production for multiple clusters for quite some time now for us.  The new twist is that now the master nodes can also run iSCSI service that listen on * and are clients to their own LFS cluster, when a transition moves the IP to a new master the iSCSI service seems to not mind at all and the iSCSI clients also do not appear to notice the change.  I have not put it under heavy loads yet but I'm hoping it will hold up well enough for my needs.  Then I can eliminate some expensive quarkie vendor iSCSI hardware or some SPOF NFS which are serving these storage role now.  The types of files being stored are actually the kind that LizardFS is best at handling.

**Comment by syadnom** [comment]
What are you using for iscsi?  how are you backing iscsi?   files on an lfs share?  I dont have a lfs cluster up right now, still testing/reading/etc.

**Comment by 4Dolio** [comment]
.oO( Digging up notes... )
Yes, the iscsi is backed by LizardFS. The master nodes are all clients of themselves.
debian packages: iscsitarget iscsitarget-source iscsitarget-dkms
Create a big sparse container file in the lfs mount with dd and then setup the iscsitarget service at all masters identically to use that file. (http://dalaris.com/how-to-configure-iscsi-target-and-iscsi-initiator-using-ubuntu-14-04-1/ was helpful)

**Comment by 4Dolio** [comment]
My notes say that I did 200 successful LFS Master transitions with no problems. And also did 2 VM migrations between Xen hosts using this SR during those master transitions.  The VM experienced about 2~4 seconds of pause during the transitions.  The Bonnie++ test results were not very good (I would not use this for DataBases for example), but this is a very round about way of providing SRs to Xen from Xen VMs using Xen Local Storage via LFS+iSCSI so.. (shrugs) But it theory (and perhaps practice) it should provide auto failover of self hosted VM storage.  It should be applicable to any iSCSI compatible client I would think.

**Comment by syadnom** [comment]
individual files per vm or one big huge one to all vm disks?  I'm thinking that just hitting localhost for iscsi might handle it, no need to use corosync or pacemaker/heartbeat/yadayada.

is lfs known to be slow for mid-file writes? ie, databases are slow... ?

**Comment by 4Dolio** [comment]
Just one huge file from the iSCSI perspective.

The way LizardFS works only one master is the acting master for the cluster at any given moment, other masters run in shadow mode and can take over at any moment. so corosync and the metadataserver OCF Script take care or promoting/demoting and moving the FloatingIP to whichever master node is in charge.  So a master node could not refer to itself as localhost because it may only be a shadow.  There is no reason you would be required to use pacemaker, but I believe you would need to use a floating IP system of some sort and it would need to promote/demote/restart master services between shadow/master modes.

I believe the MooseFS parent of LizardFS went the other route whereby masters have static addresses and you move DNS pointers and perhaps shadows/slaves can forward queues to whom ever the master happens to be.  I believe that short of a active-active+multi-master architecture which neither projects do that moving the IP arround is better than moving a DNS pointer around (ARP>DNS). I run a LFS cluster on some RaspberryPi, BananaPi, ODroids at home for fun and none of them have any names at all, so... (shrugs)

**Comment by syadnom** [comment]
maybe I said it wrong.  the iscsi initiator would point to localhost, iscsi target running on localhost, iscsi volume file sitting on lfs mount.

**Comment by 4Dolio** [comment]
OIC, that might also work, not sure if you would encounter locking issues with multiple lfs clients attempting to do simultaneous writes to the same file.  Worth a try I suppose.

The way I did it the corosync+LizardFS must already exist for the LFS cluster to handle failover and the iSCSI target service just sort of piggybacks on the same IP that the LFS Master node holds with no additional changes needed to the corosync clusters configuration to accommodate the iSCSI service.

**Comment by Zorlin** [comment]
I'm fascinated and impressed that you got this working. I'm a little
jealous - I wanted to get something similar set up in my OpenStack &
LizardFS Lab... but I haven't finished building it ...

**Comment by syadnom** [comment]
yeah, iscsi might have an issue with multiple vm's hitting the file :/  What I'm getting is that you are exporting the iscsi target *from the master* and only from the master..

I think I might go another route and export an lfs mountpoint via NFS and hit that via localhost.  NFS naturally avoids file locks (by having different files per VM) and the vm environment will handle 'split brain' to the storage..  I've had really good luck with NFS for vm storage...  The alternative is probably creating dedicated iscsi files/targets for every VM so there is no concurrency issues, and still target localhost...

**Comment by 4Dolio** [comment]
I was very sad at first when I began with this project with NFS instead of iSCSI, NFS did not work so well.  But when I switched to iSCSI it all just began to work.  It has got me thinking of new novel ways to leverage LFS+iSCSI (toss in some ZFS perhaps) in other ways to take advantage of what LFS excels at which is big files while masking what it is not so good at which is billions of small files.  Keep in mind that I have not yet gone to scale with this setup but I'm hoping it will prove more robust than our aging vendor iSCSI storage devices.

**Comment by 4Dolio** [comment]
Yes I am exporting iscsi target from the master although the iscsi service (at each master) is not bounced during a master transition.  Thus there is no need for additional corosync complexity in dealing with the iscsi service, it just seems to just work.

...But when I was using NFS I was attempting to float the IP between nodes which the NFS service did not deal with very elegantly... Your method may work better, in fact I might try it when I have a moment.

**Comment by syadnom** [comment]
if you setup iscsi for shared access, they it really shouldn't have any issues anyway because each VM should only be addressing it's own blocks (it's iscsi after all, block storage).  In theory, could host iscsi right on the vm host and run lfs client there  vm>iscsi>file on lfs>vmhost and iscsi *should* be fine.  setting it as shared removes the file lock.

I used to run a rather large xenserver cluster (full rack of r710 dual hex cores) against a rather large 60TB ZFS pool exporting iscsi as a few SR. iscsi can be tuned quite well to this kind of system.

**Comment by 4Dolio** [comment]
This was my first project setting up iSCSI target services (aside from vendor hardware) and I only went as far as necessary to get this running well enough to begin testing it.  In my case with XenServer being the initiator I don't think 'shared access' would help since XenServer only initiates to a single IP (Perhaps multi-path iSCSI? Is that a thing?)  But yes, there is a bottleneck in my setup with a single iSCSI_Target+LFS_Master IP/connection.  When there is likely to be many LFS masters which are each LFS Clients and would each by default (per LFS Topology) prefer their own Local LFS ChunkService and could in theory take advantage of parallel read access to the chunk data stored in on each LFS CS. Hummm...

**Comment by Zorlin** [comment]
@4Dolio As I understand it, MooseFS' (commercial) failover solution involves listing all eligible masters in a single A record (similar to RR DNS), and the servers handle elections and transitions among themselves. When a client connects, it queries all 3, and (I think) when a master transitions it informs all connected clients.

Haven't yet used MooseFS Pro personally, so I don't know how accurate that is.

**Comment by syadnom** [comment]
"XenServer being the initiator I don't think 'shared access' would help since XenServer only initiates to a single IP"

If you have 2 XenServer hosts accessing the same sparsefile...  They would access different blocks within the iscsi volume so wouldn't compete, but would need 'shared' access to the volume...  The real question is how will having multiple iscsi targets seeing the same 'shared' media react...  needs testing.  I have done some testing along these lines, running iscsi in kvm machines and cluster lvm (iscsi backed) and it works nicely.  iscsi just assumes that each initiator will only write to blocks it 'claims'.  if a vm is migrated, then it will access the blocks from it's new host but *not* from it's old host so all is well.

NFS works very happily in this type of setup.  I've used it extensively with glusterfs exporting nfs shares (via built-in mechanism) on localhost.  adding an NFS SR as localhost:/glusterfs/share works perfectly for live migrations etc. This also works happily with proxmox, except that proxmox has a native interface for glusterfs.  Biggest problem with XenServer is being stuck back on CentOS 6.5 :/  That and I don't really care for xen, KVM is where it's at...

**Comment by Blackpaw** [comment]
> A related question, can lizardfs present as iscsi or nfs storage? To utilize a filesystem for live migration, I either need support in proxmox (so it knows the storage is available from other nodes) or to by a dns addressed nfs or iscsi share.

A bit late, but with proxmox if you mount lizardfs on the same directory for each node (e.g /mnt/lizardfs) and set it up as shared directory storage in proxmox, you can live migrate between nodes, no need for nfs or iscsi.
